pkg/scheduler/framework/v1alpha1/framework.go

```diff
@@ -0,0 +1,117 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"encoding/json"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/klog"
+	internalcache "k8s.io/kubernetes/pkg/scheduler/internal/cache"
+)
+
+// Framework is the component responsible for initializing and
+type Framework struct {
+	registry       Registry
+	pluginContext  *PluginContext
+	cache          *ReadOnlyCache
+	plugins        map[string]Plugin // All plugins
+	reservePlugins []ReservePlugin
+	prebindPlugins []PrebindPlugin
+}
+
+var _ = FrameworkInterface(&Framework{})
+
+// PluginContext returns a pointer to PluginContext that plugins can use to
+// store their temporary data.
+func (f *Framework) PluginContext() *PluginContext {
+	return f.pluginContext
+}
+
+// SchedulerCache returns a pointer to a custom ready-only interface of the
+// scheduler cache.
+func (f *Framework) SchedulerCache() SchedulerCache {
+	return f.cache
+}
+
+// ReservePlugins returns a slice of "reserve" plugins.
+func (f *Framework) ReservePlugins() []ReservePlugin {
+	return f.reservePlugins
+}
+
+// PrebindPlugins returns a slice of "prebind" plugins.
+func (f *Framework) PrebindPlugins() []PrebindPlugin {
+	return f.prebindPlugins
+}
+
+// NewFramework initializes plugins given the configuration and the registry.
+func NewFramework(r Registry, _ json.RawMessage, c internalcache.Cache) *Framework {
+	f := &Framework{
+		registry:      r,
+		pluginContext: NewPluginContext(),
+		cache:         &ReadOnlyCache{cache: c},
+		plugins:       make(map[string]Plugin),
+	}
+
+	// TODO: The framework needs to read the scheduler config and initialize only
+	// needed plugins. In this initial version of the code, we initialize all.
+	for name, factory := range r {
+		// TODO: 'nil' should be replaced by plugin config.
+		p, err := factory(nil, f)
+		if err != nil {
+			klog.Errorf("Error initializing plugin %v: %v", name, err)
+			return nil
+		}
+		f.plugins[name] = p
+
+		// TODO: For now, we add assume any plugins that implements an extension
+		// point wants to be called at that extension point. We should change this
+		// later and add these plugins based on the configuration.
+		if rp, ok := p.(ReservePlugin); ok {
+			f.reservePlugins = append(f.reservePlugins, rp)
+		}
+		if pp, ok := p.(PrebindPlugin); ok {
+			f.prebindPlugins = append(f.prebindPlugins, pp)
+		}
+	}
+	return f
+}
+
+// ReadOnlyCache provides read-only access to some information in the scheduler
+// cache.
+type ReadOnlyCache struct {
+	cache internalcache.Cache
+}
+
+var _ = SchedulerCache(ReadOnlyCache{})
```

Comment:
Are these assertions needed? We assign a `ReadOnlyCache` value to a struct field of type `framework.SchedulerCache`

Reply:
If we fail to implement the interface, scheduler won't compile. Even without these assertions.

Reply:
That's correct, but these also provide information to the reader that these structs implement certain interfaces.